extension/httpforwarder/extension_test.go

```diff
@@ -0,0 +1,250 @@
+			expectedBackendResponseBody: []byte("\n"),
+			requestErrorAtForwarder:     true,
+			clientRequestArgs: clientRequestArgs{
+				method: "GET",
+				url:    fmt.Sprintf("http://%s/api/dosomething", listenAt),
+			},
+		},
+		{
+			name: "Error on Startup",
+			config: &Config{
+				Ingress: confighttp.HTTPServerSettings{
+					Endpoint: "invalid", // to mock error setting up listener.
+				},
+			},
+			startUpError: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if test.httpErrorFromBackend {
+					http.Error(w, "", http.StatusInternalServerError)
+					return
+				}
+
+				assert.Equal(t, getParsedURL(t, test.clientRequestArgs.url).RequestURI(), r.RequestURI)
+				assert.Equal(t, test.clientRequestArgs.method, r.Method)
+				assert.Equal(t, test.clientRequestArgs.body, string(readBody(r.Body)))
+
+				// Assert headers originating from client.
+				for k, v := range test.clientRequestArgs.headers {
+					got := r.Header.Get(k)
+					assert.Equal(t, v, got)
+				}
+
+				// Assert additional headers added by forwarder.
+				for k, v := range test.config.Egress.Headers {
+					got := r.Header.Get(k)
+					assert.Equal(t, v, got)
+				}
+
+				// Assert Via header added by the forwarder on all requests.
+				assert.Equal(t, fmt.Sprintf("%s %s", r.Proto, listenAt), r.Header.Get("Via"))
+
+				for k, v := range test.expectedHeaders {
+					w.Header().Set(k, v)
+				}
+				w.WriteHeader(test.expectedbackendStatusCode)
+				w.Write(test.expectedBackendResponseBody)
+			}))
+			defer backend.Close()
+
+			// Fill in final destination URL.
+			backendURL, _ := url.Parse(backend.URL)
+			test.config.Egress.Endpoint = backendURL.String()
+
+			// Setup forwarder with wrong final address to mock failures.
+			if test.requestErrorAtForwarder {
+				test.config.Egress.Endpoint = "http://" + testutil.GetAvailableLocalAddress(t)
+			}
+
+			hf, err := newHTTPForwarder(test.config, zap.NewNop())
+			require.NoError(t, err)
+
+			ctx := context.Background()
+			if test.startUpError {
+				require.Error(t, hf.Start(ctx, componenttest.NewNopHost()))
+				return
+			}
+			require.NoError(t, hf.Start(ctx, componenttest.NewNopHost()))
+
+			// Mock a client trying to talk to backend using the forwarder.
+			httpClient := http.Client{}
+
+			// Assert responses received by client.
+			response, err := httpClient.Do(httpRequest(t, test.clientRequestArgs))
+			require.NoError(t, err)
+			require.NotNil(t, response)
+			defer response.Body.Close()
+
+			assert.Equal(t, test.expectedbackendStatusCode, response.StatusCode)
+			if !test.requestErrorAtForwarder {
+				assert.Equal(t, string(test.expectedBackendResponseBody), string(readBody(response.Body)))
+				assert.Equal(t, fmt.Sprintf("%s %s", response.Proto, listenAt), response.Header.Get("Via"))
+			}
+
+			// Assert headers from target exist in response.
```

Comment:
would it make sense to confirm all response headers are expected here instead of just the response set ones and the earlier via check?  Otherwise something else could be unintentionally introduced and uncaught.

Reply:
I intentionally checked only the expected headers since there could be headers such as `Content-Length`, `Content-Type` and `Date` that are added by default. But if you feel strongly about testing for all headers, I can update.

Reply:
Not a blocker, but I prefer tests with complete payload verification so that any content changes are detected.

Reply:
Updated to go through all headers in response to catch any unexpected headers.